src/autograd.cpp

```diff
@@ -57,145 +57,82 @@ bool cpp_tensor_requires_grad (torch::Tensor self) {
   return lantern_Tensor_requires_grad(self.get());
 }
 
-std::deque<std::packaged_task<void*()>> tasks;
-std::deque<std::packaged_task<void()>> backward_tasks;
-std::atomic<int> n_tasks = {0};
-std::mutex tasks_mutex;
-std::mutex backward_tasks_mutex;
+namespace {
 
-void event_loop_thread(std::atomic<bool> &event_loop_running)
-{
-  
-  while (event_loop_running) {
-    
-    {
-      std::unique_lock<std::mutex> lock(tasks_mutex);
-      while (!tasks.empty()) {
-        auto task(std::move(tasks.front()));
-        tasks.pop_front();
-        
-        lock.unlock();
-        task();
-        lock.lock();
-      }
-    }
-    
+EventLoop<void*> gTasks;
+EventLoop<void> gBackwardTasks;
+
+void schedule_backward_task(std::packaged_task<void()>&& task) {
+  if (std::this_thread::get_id() == main_thread_id()) {
```

Comment:
Another simplification: using this instead of `n_tasks` to decide when to launch something in a new thread and when to continue in the current thread.